### ISSUE TITLE
Test/Solution: grocy 2.6.1 and subdirectory installation on Apache

### ISSUE THREAD

**Issue by berrnd** [grocy/grocy#694]
I've finally made some own test. It works, like this:

References #623, #619 and #661.

### Base install
```
apt install apache2 php7.3 php7.3-sqlite3 unzip
a2enmod rewrite
cd /var/www/html
wget https://releases.grocy.info/v?2.6.1 -O grocy.zip
unzip grocy.zip -d grocy
chown -R www-data:www-data ./grocy/data
```

=> Webserver document root = `/var/www/html`
=> grocy directory = `/var/www/html/grocy`

### Without URL rewriting

Changed `config.php` settings:
```
Setting('BASE_PATH', '/grocy/public/index.php');
Setting('BASE_URL', 'http://<HOST>/grocy/public');
Setting('DISABLE_URL_REWRITING', true);
```

_(Nothing else was changed in the default webserver/PHP config files.)_

=> `http://<HOST>/grocy/public/index.php/` is working (note the trailing slash for the root/main route)

### With URL rewriting

Changed `config.php` settings
```
Setting('BASE_PATH', '/grocy/public');
Setting('BASE_URL', 'http://<HOST>/grocy/public');
```

To allow `.htaccess` overrides, this was added to `/etc/apache2/apache2.conf`:
```
<Directory /var/www/html/grocy>
	AllowOverride All
</Directory>
```

=> `http://<HOST>/grocy/public/` is working (note the trailing slash for the root/main route)

#### Additional notes

Add a `.htaccess` file in `/var/www/html/grocy/data` with the following content to disallow direct access to the data dirctory (e. g. that someone can download the grocy database file)
```
Deny from all
```

## Conclusion
Basically, `BASE_PATH` needs to be set to the part (of the URL) after the document root, if URL rewriting is disabled, including `index.php`.

**Comment by TaciteOFF** [comment]
Still doesn't work for me.

> File: /srv/disk18/3376730/www/mysubdomain/vendor/slim/slim/Slim/Middleware/RoutingMiddleware.php

**Comment by berrnd** [comment]
What's your webserver document root? What's your grocy `BASE_PATH` setting?

If you're in a shared hosting environment, maybe somewhere the document root is mentioned or try to get it by PHP (create any .php file with `<?php echo $_SERVER['DOCUMENT_ROOT']; ?>` in it).

**Comment by TaciteOFF** [comment]
DOCUMENT_ROOT gives /home/www/mysubdomain.tld

So I changed the config.php file to

Setting('BASE_PATH', '/home/www/mysubdomain.tld/public/index.php');
Setting('BASE_URL', 'http://mysubdomain.tld/public');

And same error

**Comment by berrnd** [comment]
Please read my example above and the comment in `config-php.dist`, `BASE_PATH` needs to be relative to the document root, so probably (if URL rewriting is working):
```
Setting('BASE_PATH', '/public');
Setting('BASE_URL', 'http://mysubdomain.tld/public');
```
=> `http://mysubdomain.tld/public/stockoverview`

Or probably if URL rewriting is not working:
```
Setting('BASE_PATH', '/public/index.php');
Setting('BASE_URL', 'http://mysubdomain.tld/public');
Setting('DISABLE_URL_REWRITING', true);
```
=> `http://mysubdomain.tld/public/index.php/stockoverview`

**Comment by TaciteOFF** [comment]
Thanks!

Now I have a different error.

`Warning: Use of undefined constant GROCY_USER_ID - assumed 'GROCY_USER_ID' (this will throw an Error in a future version of PHP) in /srv/disk18/3376730/www/mysubdomain.tld/controllers/StockController.php on line 16
`

> Message: SQLSTATE[HY000]: General error: 1 no such table: user_settings

**Comment by berrnd** [comment]
Can only happen if migrations went wrong (if it's a new installation, just restart by deleting `data/grocy.db` and then visit the root route (= `http://mysubdomain.tld/public/index.php` or `http://mysubdomain.tld/public` depending on URL rewriting) to execute database migrations), or if `MODE` is not `production`.

**Comment by TaciteOFF** [comment]
It is a new installation. I deleted grocy.db
When I go to index.php I get a 404 error on RoutingMiddleware.php :/
Sorry I can't help more than that.

**Comment by berrnd** [comment]
However, in https://github.com/grocy/grocy-desktop/issues/11#issuecomment-605407639 is a fully populated empty v.2.6.1 `grocy.db` - just replace that (and make sure te permissions are set correctly), maybe that helps.

**Comment by TaciteOFF** [comment]
After logging in with admin:admin, everything seems to work :-)

**Comment by berrnd** [comment]
Can you do a quick test again?

Delete again `data/grocy.db`, then visit again `http://mysubdomain.tld/public/index.php/` (note the trailling slash, that's important). Does that work? I was able to get it to work locally by that just now...

**Comment by TaciteOFF** [comment]
Yup it works. 
I still have to log-in tho, because before first log-in I have :

Warning: Use of undefined constant GROCY_USER_ID - assumed 'GROCY_USER_ID'

**Comment by emk2203** [comment]
Just for clarification: When you write that `http://<HOST>/grocy/public/index.php/` is working, you are referring to the `config.php` entry, the webserver is still at `http://<HOST>` when you enter the address in your client device?

**Comment by berrnd** [comment]
Yes, if I understand what you mean.

By the example above:
`http://<HOST>` serves `/var/html`.
grocy is in the subdirectory `/var/html/grocy` which is served at `http://<HOST>/grocy` so the mentioned `http://<HOST>/grocy/public/index.php/` for a fully workable grocy path (main/root route) (with disabled URL rewriting).

**Comment by berrnd** [comment]
Basically, `BASE_PATH` needs to be set to the part (of the URL) after your document root, if URL rewriting is disabled, including `index.php`.

`http://<HOST>/grocy/public/index.php/` => `BASE_PATH` `/grocy/public/index.php`

Does it work for you now @emk2203?

**Comment by emk2203** [comment]
Not quite, I need to see what to adapt for nginx here. My first quick try was not successful. I edit this later when I find out something.

**Comment by zmijunkie** [comment]
here my changes in config.php using nginx:
```
Setting('BASE_PATH', '/grocy/public');
Setting('BASE_URL', 'http://rucola.local/grocy/public');
```
This website changed everything for me:
https://serversforhackers.com/c/nginx-php-in-subdirectory
after watching this video I made my Raspberry4 serving grocy

```server {
        listen 80;
        root /var/www;
        index index.php index.html index.htm;

	location = /favicon.ico {
		log_not_found off;
		access_log off;
	}

        # Use empty location blocks to avoid the redirect for certain requests
        location ~ \.js$ {}
        location ~ \.css$ {}
        location ~ \.woff$ {}
        location ~ \.woff2$ {}
        location ~ \.svg$ {}
        location ~ \.map$ {}

        location /grocy  {
           alias /var/www/grocy/public;
           try_files $uri $uri/ @grocy;

            location ~ \.php$ {
                include snippets/fastcgi-php.conf;
                fastcgi_param SCRIPT_FILENAME $request_filename;
                fastcgi_pass unix:/var/run/php/php7.3-fpm.sock;
            }
        }

	location @grocy  {
           rewrite /(.*)$ /grocy/index.php?/$1 last;
        }

        location ~ \.php$ {
                include snippets/fastcgi-php.conf;
                fastcgi_pass unix:/var/run/php/php7.3-fpm.sock;
                # fastcgi_index index.php;
                # include fastcgi_params;
        }
}
```

**Comment by berrnd** [comment]
@zmijunkie thank you a lot! :)

So we have a working install-in-a-subdirectory-example for Apache and nginx, I think this can be considered resolved (in fact there was never an issue, it was just unclear how to configure everything properly).

I will add a reference to https://grocy.info/links and the RAEDME before closing this.

**Comment by kriddles** [comment]
> Setting('BASE_PATH', '/grocy/public');
> Setting('BASE_URL', 'http://rucola.local/grocy/public');

I think you should be able to remove the /public from these settings.

**Comment by dragoscazaceanu1313** [comment]
noob here, love grocy but not so much into coding. would you be so kind as to point out where this stack needs to be placed/replaced? 

        listen 80;
        root /var/www;
        index index.php index.html index.htm;

	location = /favicon.ico {
		log_not_found off;
		access_log off;
	}

        # Use empty location blocks to avoid the redirect for certain requests
        location ~ \.js$ {}
        location ~ \.css$ {}
        location ~ \.woff$ {}
        location ~ \.woff2$ {}
        location ~ \.svg$ {}
        location ~ \.map$ {}

        location /grocy  {
           alias /var/www/grocy/public;
           try_files $uri $uri/ @grocy;

            location ~ \.php$ {
                include snippets/fastcgi-php.conf;
                fastcgi_param SCRIPT_FILENAME $request_filename;
                fastcgi_pass unix:/var/run/php/php7.3-fpm.sock;
            }
        }

	location @grocy  {
           rewrite /(.*)$ /grocy/index.php?/$1 last;
        }

        location ~ \.php$ {
                include snippets/fastcgi-php.conf;
                fastcgi_pass unix:/var/run/php/php7.3-fpm.sock;
                # fastcgi_index index.php;
                # include fastcgi_params;
        }
}

i've managed to install using the instructions here https://github.com/grocy/grocy/issues/649 so i'm definitely using nginx, but not sure what to edit. 

Also, a interesting fact, maybe bernd can shed some light on this. After installing according to the rules in 649 everything was working fine albeit with an empty db. when i try to change the db by using the one in the windows installation (where i have all product and pictures configured), the folder structure does not change ? so why are these steps requred ? i don't see any extra nesting ocurring here then what was happening initially. 

What i tried:
1. after correct and running installation as per 649, my installation is currently under /var/www/html
2. i created a folder in data called old_data where i moved plugins, viewcache, config.php and grocy.db
3. i created a folder called new_data where i copied the files from my running windows installation: plugins, storage viewcache, config.php, grocy.db, without the override folder.
4. i moved the windows installations files from /var/www/html/data/new_data /var/www/html/data/grocy
5. i edited config.php to look like this:

<?php

# Settings can also be overwritten in two ways
#
# First priority
# A .txt file with the same name as the setting in /data/settingoverrides
# the content of the file is used as the setting value
#
# Second priority
# An environment variable with the same name as the setting and prefix "GROCY_"
# so for example "GROCY_BASE_URL"
#
# Third priority
# The settings defined here below


# Either "production", "dev", "demo" or "prerelease"
# When not "production", authentication will be disabled and
# demo data will be populated during database migrations
Setting('MODE', 'production');

# Either "en" or "de" or the directory name of
# one of the other available localization folders in the "/localization" directory
Setting('CULTURE', 'en');

# This is used to define the first day of a week for calendar views in the frontend,
# leave empty to use the locale default
# Needs to be a number where Sunday = 0, Monday = 1 and so forth
Setting('CALENDAR_FIRST_DAY_OF_WEEK', '');

# If calendars should show week numbers
Setting('CALENDAR_SHOW_WEEK_OF_YEAR', true);

# To keep it simple: grocy does not handle any currency conversions,
# this here is used to format all money values,
# so doesn't really matter, but should be the
# ISO 4217 code of the currency ("USD", "EUR", "GBP", etc.)
Setting('CURRENCY', 'RON');

# When running grocy in a subdirectory, this should be set to the relative path, otherwise empty
# Example:
#  Webserver root directory = /var/www
#  grocy directory = /var/www/grocy
#  => BASE_PATH = /grocy
Setting('BASE_PATH', 'html/grocy/public');

# The base url of your installation,
# should be just "/" when running directly under the root of a (sub)domain
# or for example "https://example.com/grocy" when using a subdirectory
Setting('BASE_URL', 'http://102.168.0.111/html/grocy/public');

# The plugin to use for external barcode lookups,
# must be the filename without .php extension and must be located in /data/plugins,
# see /data/plugins/DemoBarcodeLookupPlugin.php for an example implementation
Setting('STOCK_BARCODE_LOOKUP_PLUGIN', 'DemoBarcodeLookupPlugin');

# If, however, your webserver does not support URL rewriting, set this to true
Setting('DISABLE_URL_REWRITING', false);

# Specify an custom homepage if desired - by default the homepage will be set to the stock overview,
# this needs to be one of the following values:
# stock, shoppinglist, recipes, chores, tasks, batteries, equipment, calendar, mealplan
Setting('ENTRY_PAGE', 'stock');

# Set this to true if you want to disable authentication / the login screen,
# places where user context is needed will then use the default (first existing) user
Setting('DISABLE_AUTH', false);

# Set this to true if you want to disable the ability to scan a barcode via the device camera (Browser API)
Setting('DISABLE_BROWSER_BARCODE_CAMERA_SCANNING', false);

# Set this if you want to have a different start day for the weekly meal plan view,
# leave empty to use CALENDAR_FIRST_DAY_OF_WEEK (see above)
# Needs to be a number where Sunday = 0, Monday = 1 and so forth
Setting('MEAL_PLAN_FIRST_DAY_OF_WEEK', '');


# Default user settings
# These settings can be changed per user, here the defaults
# are defined which are used when the user has not changed the setting so far

# Night mode related
DefaultUserSetting('night_mode_enabled', false); // If night mode is enabled always
DefaultUserSetting('auto_night_mode_enabled', false); // If night mode is enabled automatically when inside a given time range (see the two settings below)
DefaultUserSetting('auto_night_mode_time_range_from', "20:00"); // Format HH:mm
DefaultUserSetting('auto_night_mode_time_range_to', "07:00"); // Format HH:mm
DefaultUserSetting('auto_night_mode_time_range_goes_over_midnight', true); // If the time range above goes over midnight
DefaultUserSetting('currently_inside_night_mode_range', false); // If we're currently inside of night mode time range (this is not user configurable, but stored as a user setting because it's evaluated client side to be able to use the client time instead of the maybe different server time)

# Keep screen on settings
DefaultUserSetting('keep_screen_on', false); // Keep the screen always on
DefaultUserSetting('keep_screen_on_when_fullscreen_card', false); // Keep the screen on when a "fullscreen-card" is displayed

# Stock settings
DefaultUserSetting('product_presets_location_id', -1); // Default location id for new products (-1 means no location is preset)
DefaultUserSetting('product_presets_product_group_id', -1); // Default product group id for new products (-1 means no product group is preset)
DefaultUserSetting('product_presets_qu_id', -1); // Default quantity unit id for new products (-1 means no quantity unit is preset)
DefaultUserSetting('stock_expring_soon_days', 5);
DefaultUserSetting('stock_default_purchase_amount', 0);
DefaultUserSetting('stock_default_consume_amount', 1);
DefaultUserSetting('scan_mode_consume_enabled', false);
DefaultUserSetting('scan_mode_purchase_enabled', false);

# Shopping list settings
DefaultUserSetting('shopping_list_to_stock_workflow_auto_submit_when_prefilled', false); // Automatically do the booking using the last price and the amount of the shopping list item, if the product has "Default best before days" set
DefaultUserSetting('shopping_list_show_calendar', false);
DefaultUserSetting('shopping_list_disable_auto_compact_view_on_mobile', false);

# Recipe settings
DefaultUserSetting('recipe_ingredients_group_by_product_group', false); // Group recipe ingredients by their product group

# Chores settings
DefaultUserSetting('chores_due_soon_days', 5);

# Batteries settings
DefaultUserSetting('batteries_due_soon_days', 5);

# Tasks settings
DefaultUserSetting('tasks_due_soon_days', 5);

# If the page should be automatically reloaded when there was
# an external change
DefaultUserSetting('auto_reload_on_db_change', true);

# Show a clock in the header next to the logo or not
DefaultUserSetting('show_clock_in_header', false);


# Feature flags
# grocy was initially about "stock management for your household", many other things
# came and still come by, because they are useful - here you can disable the parts
# which you don't need to have a less cluttered UI
# (set the setting to "false" to disable the corresponding part, which should be self explanatory)
Setting('FEATURE_FLAG_STOCK', true);
Setting('FEATURE_FLAG_SHOPPINGLIST', true);
Setting('FEATURE_FLAG_RECIPES', true);
Setting('FEATURE_FLAG_CHORES', true);
Setting('FEATURE_FLAG_TASKS', true);
Setting('FEATURE_FLAG_BATTERIES', true);
Setting('FEATURE_FLAG_EQUIPMENT', true);
Setting('FEATURE_FLAG_CALENDAR', true);


# Sub feature flags
Setting('FEATURE_FLAG_STOCK_PRICE_TRACKING', true);
Setting('FEATURE_FLAG_STOCK_LOCATION_TRACKING', true);
Setting('FEATURE_FLAG_STOCK_BEST_BEFORE_DATE_TRACKING', true);
Setting('FEATURE_FLAG_STOCK_PRODUCT_OPENED_TRACKING', true);
Setting('FEATURE_FLAG_STOCK_PRODUCT_FREEZING', true);
Setting('FEATURE_FLAG_SHOPPINGLIST_MULTIPLE_LISTS', true);
Setting('FEATURE_FLAG_CHORES_ASSIGNMENTS', true);


# Feature settings
Setting('FEATURE_SETTING_STOCK_COUNT_OPENED_PRODUCTS_AGAINST_MINIMUM_STOCK_AMOUNT', true); // When set to false, opened products will not be considered for minimum stock amounts

Now whenever i try to open http://192.168.0.111 it just shows internal 500 error. 
What am I doing wrong ? 

Thanks and sorry for the long post. :D

**Comment by berrnd** [comment]
@dragoscazaceanu1313 please use [the subbredit](https://www.reddit.com/r/grocy/) for custom setup questions, thanks.
